### PR TITLE
pfcli update: signup, org, project

### PR DESCRIPTION
**Pull request by creeper00**
`org invite`, `org set-privilege`, `project add-user`, `project set-privilege` can be used when org is set.

### CODE REVIEW

**Comment by kooyunmo** [comment]
Can you please add unittests for client services at `tests/client/`?

File: pfcli/main.py [hunk 1]
```diff
@@ -49,6 +49,25 @@
...
 )
 
+@app.command(help="Sign up to PerfiFlow")
+def signup(
+    username: str = typer.Option(..., prompt="Enter Username"),
+    name: str = typer.Option(..., prompt="Enter Name"),
+    email: str = typer.Option(..., prompt="Enter Email"),
+    password: str = typer.Option(..., prompt="Enter Password", hide_input=True)
+):
+    r = requests.post(
```
Comment (ktaebum) [inline pfcli/main.py:59]: Could you implement this code within the `UserClientService`?

File: pfcli/main.py [hunk 2]
```diff
@@ -94,3 +113,13 @@ def passwd(
     client.change_password(old_password, new_password)
 
     typer.secho("Password is changed successfully!", fg=typer.colors.BLUE)
+
+
+def _validate(_, token: str = typer.Option(..., prompt="Enter Code")):
+    r = requests.post(get_auth_uri("pf_user/self_signup/confirm"), json={"email_token": token})
```
Comment (ktaebum) [inline pfcli/main.py:119]: Could you implement this code within the `UserClientService`?

File: pfcli/group.py [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 """PeriFlow Group (Organization) CLI"""
 
+from enum import Enum
+from pfcli.service.client.user import UserClientService
```
Comment (ktaebum) [inline pfcli/group.py:6]: plz group `pfcli` imports together

File: pfcli/project.py [hunk 4]
```diff
@@ -2,8 +2,11 @@
 
 """PeriFlow Project CLI"""
 
+from enum import Enum
```
Comment (ktaebum) [inline pfcli/project.py:5]: order
```
import uuid
from enum import Enum
from typing import List, Optional, Union

from pfcli...
```

File: pfcli/main.py [hunk 5]
```diff
@@ -94,3 +113,13 @@ def passwd(
...
+    try:
+        r.raise_for_status()
+        typer.echo("\n\nVerified!")
+        typer.echo("Sign up success! Please sign in.")
+    except HTTPError:
+        secho_error_and_exit("Invalid code... Please Try again.")
```
Comment (kooyunmo) [inline pfcli/main.py:125]: `HTTPError` is not always related to invalid code, but it also related to duplicate email or username.
When duplicate email address is used, the error response code is 409 and the content is as follows:

```
'{"error":"not_unique_field","error_description":"Field should be unique"}'
```

File: pfcli/main.py [hunk 6]
```diff
@@ -49,6 +49,25 @@
     headers=["Name", "Username", "Email"]
 )
 
+@app.command(help="Sign up to PerfiFlow")
+def signup(
+    username: str = typer.Option(..., prompt="Enter Username"),
+    name: str = typer.Option(..., prompt="Enter Name"),
+    email: str = typer.Option(..., prompt="Enter Email"),
+    password: str = typer.Option(..., prompt="Enter Password", hide_input=True)
```
Comment (kooyunmo) [inline pfcli/main.py:57]: Please make it re-enter password for confirmantion.

e.g., "Enter password" -> "Enter the password again"

File: pfcli/service/client/group.py [hunk 7]
```diff
@@ -32,6 +32,27 @@ def get_group(self, pf_group_id: uuid.UUID) -> dict:
...
+        safe_request(self.post, err_prefix="Invalid code... Please Try again.")(
+            path="invite/confirm",
+            json={
+                "email_token": token
+            }
+        )
+
+    def get_user(self, pf_group_id: str, username: str) -> dict:
+        get_response_dict = safe_request(self.list, err_prefix="Failed to get user in organization")
+        return paginated_get(get_response_dict, path=f"{pf_group_id}/pf_user", **{ "search": username })
```
Comment (kooyunmo) [inline pfcli/service/client/group.py:54]:
```suggestion
        return paginated_get(get_response_dict, path=f"{pf_group_id}/pf_user", search=username)
```

File: pfcli/service/client/user.py [hunk 8]
```diff
@@ -24,6 +24,41 @@ def change_password(self, old_password: str, new_password: str) -> None:
...
+        safe_request(self.partial_update, err_prefix="Failed to update privilege level in group")(
+            pk=pf_user_id,
+            path=f"pf_group/{pf_group_id}/privilege_level",
+            json={
+                "privilege_level": privilege_level
+            }
+        )
+
+    # Project
+    def get_project_membership(self, pf_project_id):
```
Comment (kooyunmo) [inline pfcli/service/client/user.py:38]:
```suggestion
    def get_project_membership(self, pf_project_id: str) -> dict:
```

File: pfcli/service/client/user.py [hunk 9]
```diff
@@ -24,6 +24,41 @@ def change_password(self, old_password: str, new_password: str) -> None:
...
+        response = safe_request(self.retrieve, err_prefix="Failed identify member in project")(
+            pk=self.user_id,
+            path=f"pf_project/{pf_project_id}",
+        )
+        return response.json()
+
+    def add_to_project(self, pf_user_id: str, pf_project_id: str, access_level: str):
```
Comment (kooyunmo) [inline pfcli/service/client/user.py:45]:
```suggestion
    def add_to_project(self, pf_user_id: str, pf_project_id: str, access_level: str) -> None:
```

File: pfcli/service/client/user.py [hunk 10]
```diff
@@ -24,6 +24,41 @@ def change_password(self, old_password: str, new_password: str) -> None:
...
+        safe_request(self.post, err_prefix="Failed to add user to project")(
+            path=f"{pf_user_id}/pf_project/{pf_project_id}",
+            json={
+                "access_level": access_level
+            }
+        )
+
+    def set_project_privilege(self, pf_user_id: str, pf_project_id: str, access_level: str):
```
Comment (kooyunmo) [inline pfcli/service/client/user.py:53]:
```suggestion
    def set_project_privilege(self, pf_user_id: str, pf_project_id: str, access_level: str) -> None:
```

File: pfcli/project.py [hunk 11]
```diff
@@ -134,3 +143,79 @@ def delete(
...
+    group_project_client: GroupProjectClientService = build_client(ServiceType.GROUP_PROJECT)
+    projects = group_project_client.list_projects()
+    for project in projects:
+        if project['name'] == name:
+            return project['id']
+
+    secho_error_and_exit(f"No project exists with name {name}.")
+
+
+def _check_project_and_get_id(name: str) -> tuple[str,str]:
```
Comment (kooyunmo) [inline pfcli/project.py:206]: import `Tuple` from `typing` and

```suggestion
def _check_project_and_get_id(name: str) -> Tuple[str, str]:
```

File: pfcli/project.py [hunk 12]
```diff
@@ -134,3 +143,79 @@ def delete(
...
+    """Get org_id and project_id if valid"""
+
+    user_client: UserClientService = build_client(ServiceType.USER)
+
+    org = _get_current_org()
+    project_id = _get_project_id_by_name(name)
+
+    if org['privilege_level'] == 'owner':
```
Comment (kooyunmo) [inline pfcli/project.py:214]: Better to use the defined enum

```suggestion
    if org['privilege_level'] == GroupAccessLevel.OWNER:
```

File: pfcli/project.py [hunk 13]
```diff
@@ -134,3 +143,79 @@ def delete(
...
+        return org['id'], project_id
+
+    requester = user_client.get_project_membership(project_id)
+    if requester['access_level'] != 'admin':
```
Comment (kooyunmo) [inline pfcli/project.py:218]:
```suggestion
    if requester['access_level'] != ProjectAccessLevel.ADMIN:
```

File: pfcli/project.py [hunk 14]
```diff
@@ -35,6 +38,12 @@
     headers=['Organization ID', 'Project ID', 'Name']
 )
 
+class ProjectAccessLevel(str, Enum):
```
Comment (kooyunmo) [inline pfcli/project.py:41]:
```suggestion

class ProjectAccessLevel(str, Enum):
```

**Review by kooyunmo — CHANGES_REQUESTED** [review]

File: pfcli/main.py (hunk 1 above)
Comment (creeper00) [reply pfcli/main.py:59]: When implemented in UserClientService, an access token is required

Comment (creeper00) [reply pfcli/main.py:59]: Oh, please wait I'll check again

Comment (creeper00) [reply pfcli/main.py:59]: Yes, `auto_token_refresh` is called when initializing `UserClientService`. Shall I define a new class like `UserClientAuthService`?

File: pfcli/project.py [hunk 15]
```diff
@@ -134,3 +143,79 @@ def delete(
...
+
+    org_id, project_id = _check_project_and_get_id(name)
+    user_id = _get_org_user_id_by_name(org_id, username)
+
+    user_client.add_to_project(user_id, project_id, access_level)
+    typer.secho(f"User successfully added to {name}")
+
+
+@app.command("set-privilege", help="set privilege level")
+def set_privilege(
```
Comment (kooyunmo) [inline pfcli/project.py:173]: Similar to the comment above, I want the project name to be configured from the current working project.

File: pfcli/main.py (hunk 1 above)
Comment (ktaebum) [reply pfcli/main.py:59]: Or I think we can implement a function such as `bare_post` in `ClientService` which is not decorated by `auto_token_refresh`

Comment (ktaebum) [reply pfcli/main.py:59]: @kooyunmo any idea?

Comment (ktaebum) [reply pfcli/main.py:59]: I think just `SignUpService` is enough

Comment (ktaebum) [reply pfcli/main.py:59]: and we could implement such as `bare_post` in `ClientService`, which does not use auth token

Comment (ktaebum) [reply pfcli/main.py:59]:
```
class UserSignUpService(ClientService):
    def signup(self):
        safe_request(self.bare_post)

    def verify(self):
        safe_request(self.bare_post)
```
...

File: pfcli/service/client/__init__.py [hunk 16]
```diff
@@ -35,11 +35,13 @@
...
     UserClientService,
     UserGroupClientService,
-    UserGroupProjectClientService
+    UserGroupProjectClientService,
+    UserSignUpService
 )
 from pfcli.utils import get_auth_uri, get_mr_uri, get_pfs_uri, get_uri, get_wss_uri
 
 client_template_map: Dict[ServiceType, Tuple[Type[ClientService], Template]] = {
+    ServiceType.SIGNUP: (UserSignUpService, Template(get_auth_uri('pf_user'))),
```
Comment (ktaebum) [inline pfcli/service/client/__init__.py:44]: maybe we can use `pf_user/self_signup`?

File: pfcli/main.py [hunk 17]
```diff
@@ -49,6 +49,25 @@
...
+    password: str = typer.Option(..., prompt="Enter Password", hide_input=True),
+    confirm_password: str = typer.Option(..., prompt="Enter the password again (confirmation)", hide_input=True)
+):
+    if password != confirm_password:
+        secho_error_and_exit("Passwords did not match.")
+
+    client: UserSignUpService = build_client(ServiceType.SIGNUP)
+    result = client.sign_up(username, name, email, password)
+    if "error_description" in result:
+        secho_error_and_exit(f'\n\nSignup failed... {result["error_description"]}')
```
[the remark this answers is not in the document]
Comment (creeper00) [reply pfcli/main.py:66]: For some reason, safe_request is not catching error for `signup` and `verify`.
`{"error":"not_unique_field","error_description":"Field should be unique"}` is not getting caught

**Pull request merged**